### DISCUSSION TITLE
Linking fails if PyO3 is used alongside tch

### DISCUSSION THREAD

**Discussion by diptanu** [PyO3/pyo3#3186] [Questions]
Hi folks, I am trying to use PyO3 and tch-rs in a project and linking fails if PyO3 is introduced to the project with the following error -
```
   warning: `indexify` (lib) generated 7 warnings
   error: linking with `cc` failed: exit status: 1
     |
     = note: "cc" "-m64" "/tmp/rustcJhnAVi/symbols.o" "/home/diptanuc/Projects/indexify/target/debug/deps/indexify-15e165cff9bd3048.14wlvzqavayxl80f.rcgu.o" "/home/diptanuc/Projects/indexify/target/debug/deps/indexify-15e165cff9bd3048.18wb4faep18xt9lx.rcgu.o" "/home/diptanuc/Projects/indexify/target/debug/deps/indexify-15e165cff9bd3048.1aui2inlwnkujm30.rcgu.o" "/home/diptanuc/Projects/indexify/target/debug/deps/indexify-
     = note: /usr/bin/ld: /home/diptanuc/Projects/indexify/target/debug/deps/libtorch_sys-3cac1932230d9071.rlib(torch_api.o): in function `c10::Device::validate()':
             /home/diptanuc/Src/libtorch/include/c10/core/Device.h:166: undefined reference to `c10::detail::torchInternalAssertFail(char const*, char const*, unsigned int, char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)' 

     = note: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
     = note: use the `-l` flag to specify native libraries to link
     = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)
```

Any thoughts or ideas? The operating system is Linux/Ubuntu 22.04 I am assuming it's because of some linker args PyO3 passes along to rustc.

Here is a reproduction of the issue -  https://github.com/diptanu/pyo3-tch

**Comment by diptanu** [comment]
Looks like the compilation fails because PyO3's build script adds -L native=/usr/lib/x86_64-linux-gnu, If I remove this it works. Can someone explain why this is required? I am guessing it has to do something with linking with Python.

**Reply by adamreichold** [reply]
It it most likely part of the system configuration handed to us by the Python interpreter. Could you run your build with the environment variable `PYO3_PRINT_CONFIG` set? This will fail the build but also print out the information our build script has gathered.

**Reply by diptanu** [reply]
Thanks for responding, here is the config -
```
  -- PYO3_PRINT_CONFIG=1 is set, printing configuration and halting compile --
  implementation=CPython
  version=3.10
  shared=true
  abi3=false
  lib_name=python3.10
  lib_dir=/usr/lib/x86_64-linux-gnu
  executable=/home/diptanuc/Projects/pyo3test/ve/bin/python
  pointer_width=64
  build_flags=
  suppress_build_script_link_lines=false

  note: unset the PYO3_PRINT_CONFIG environment variable and retry to compile with the above config
```

**Reply by adamreichold** [reply]
As you can see, your Python installation has its `libpython3.10.so` located in `/usr/lib/x86_64-linux-gnu` and to link against it, we need to add the directory to the linker search path.

You can suppress this by creating a build configuration as described https://pyo3.rs/main/building_and_distribution#advanced-config-files, i.e. by setting `suppress_build_script_link_lines=true`. But then you need to take of linking Python manually, for example using the `extra_build_script_lines` variable.

In this particular case, I wonder why having this directory on the search path is a problem in the first place. Is there a conflicting PyTorch installation present on your system? Could `tch-rs` be instructed to use that instead of a separate one?
